File: src/main/java/gargoyle/rpycg/ui/KeyText.java
package gargoyle.rpycg.ui;

import gargoyle.rpycg.fx.FXContextFactory;
import javafx.beans.DefaultProperty;
import javafx.beans.property.Property;
import javafx.beans.property.ReadOnlyProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.scene.control.TextField;
import javafx.scene.control.Tooltip;
import javafx.scene.input.KeyCodeCombination;
import javafx.scene.input.KeyCombination;
import javafx.scene.input.KeyEvent;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.PropertyKey;

import java.util.Optional;
import java.util.ResourceBundle;

@DefaultProperty("keyCode")
public final class KeyText extends TextField {
    @PropertyKey(resourceBundle = "gargoyle.rpycg.ui.KeyText")
    private static final String LC_DEFAULT = "default-is";
    @PropertyKey(resourceBundle = "gargoyle.rpycg.ui.KeyText")
    private static final String LC_PRESS_ANY_KEY = "press-any-key";
    @NotNull
    private final Property<Boolean> activated;
    @NotNull
    private final Property<KeyCodeCombination> combination;
    @NotNull
    private final Property<KeyCodeCombination> defaultCombination;

    public KeyText() {
        defaultCombination = new SimpleObjectProperty<>(null);
        combination = new SimpleObjectProperty<>(null);
        activated = new SimpleBooleanProperty(false);
        Optional<ResourceBundle> optional = FXContextFactory.currentContext().loadResources(KeyText.class);
        String tooltip = optional.map(resources -> resources.getString(LC_DEFAULT)).orElse(LC_DEFAULT);
        String pressKey = optional.map(resources -> resources.getString(LC_PRESS_ANY_KEY)).orElse(LC_PRESS_ANY_KEY);
        defaultCombination.addListener((observable, oldValue, newValue) -> {
            if (oldValue == null) {
                combination.setValue(newValue);
                updateTooltip(newValue, tooltip);
            }
            updateTooltip(newValue, tooltip);
        });
        updateTooltip(defaultCombination.getValue(), tooltip);
        combination.addListener((observable, oldValue, newValue) -> {
            if (oldValue == null) {
                updateTooltip(newValue, tooltip);
            }
            setText(newValue.getDisplayText());
        });
        activated.addListener((observable, oldValue, newValue) -> {
            if (newValue) {
                setText(pressKey);
            } else {
                setText(combination.getValue().getDisplayText());
            }
        });
        setEditable(false);
        setOnKeyReleased(event -> {
            if (activated.getValue()) {
                if (event.getCode().isLetterKey() || event.getCode().isDigitKey()) {
                    combination.setValue(toCombination(event));
                }
                activated.setValue(false);
            }
        });
        focusedProperty().addListener((observable, oldValue, newValue) -> activated.setValue(newValue));
        setOnMouseClicked(event -> {
            requestFocus();
            activated.setValue(true);
        });
    }

    private void updateTooltip(@Nullable KeyCodeCombination keyCodeCombination, @NotNull String tooltipString) {
        Optional.ofNullable(keyCodeCombination).ifPresent(newKeyCode ->
                setTooltip(new Tooltip(tooltipString + ' ' + newKeyCode.getDisplayText())));
    }

    @NotNull
    private static KeyCodeCombination toCombination(@NotNull KeyEvent event) {
        return new KeyCodeCombination(event.getCode(),
                event.isShiftDown() ? KeyCombination.ModifierValue.DOWN : KeyCombination.ModifierValue.UP,
                event.isControlDown() ? KeyCombination.ModifierValue.DOWN : KeyCombination.ModifierValue.UP,
                event.isAltDown() ? KeyCombination.ModifierValue.DOWN : KeyCombination.ModifierValue.UP,
                event.isMetaDown() ? KeyCombination.ModifierValue.DOWN : KeyCombination.ModifierValue.UP,
                event.isShortcutDown() ? KeyCombination.ModifierValue.DOWN : KeyCombination.ModifierValue.UP);
    }

    @NotNull
    public ReadOnlyProperty<Boolean> activatedProperty() {
        return activated;
    }

    @NotNull
    public ReadOnlyProperty<KeyCodeCombination> combinationProperty() {
        return combination;
    }

    @NotNull
    public Property<KeyCodeCombination> defaultCombinationProperty() {
        return defaultCombination;
    }

    @NotNull
    public Boolean getActivated() {
        return activated.getValue();
    }

    @NotNull
    public KeyCodeCombination getCombination() {
        return combination.getValue();
    }

    @NotNull
    public KeyCodeCombination getDefaultCombination() {
        return defaultCombination.getValue();
    }

    public void setDefaultCombination(@NotNull KeyCodeCombination defaultCombination) {
        this.defaultCombination.setValue(defaultCombination);
    }

    public void reset() {
        combination.setValue(defaultCombination.getValue());
        activated.setValue(false);
    }
}

File: app/src/main/kotlin/gargoyle/rpycg/ui/KeyText.kt
package gargoyle.rpycg.ui

import gargoyle.fx.FxContext
import gargoyle.fx.FxRun
import gargoyle.fx.FxUtil.get
import javafx.beans.DefaultProperty
import javafx.beans.property.Property
import javafx.beans.property.ReadOnlyProperty
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.value.ObservableValue
import javafx.event.EventHandler
import javafx.scene.control.TextField
import javafx.scene.control.Tooltip
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyCodeCombination
import javafx.scene.input.KeyCombination
import javafx.scene.input.KeyEvent
import java.util.*

@DefaultProperty("keyCode")
class KeyText : TextField() {
    private val activated: Property<Boolean> = SimpleBooleanProperty(false)
    private val combination: Property<KeyCodeCombination> = SimpleObjectProperty(NONE)
    private val defaultCombination: Property<KeyCodeCombination> = SimpleObjectProperty(NONE)

    init {
        val resources: ResourceBundle = FxContext.current.loadResources(KeyText::class)!!
        val tooltip = resources[LC_DEFAULT, LC_DEFAULT]
        val pressKey = resources[LC_PRESS_ANY_KEY, LC_PRESS_ANY_KEY]
        defaultCombination.addListener { _: ObservableValue<out KeyCodeCombination>, oldValue: KeyCodeCombination?, newValue: KeyCodeCombination ->
            if (null == oldValue || NONE == oldValue) combination.value = newValue
            updateTooltip(newValue, tooltip)
        }
        updateTooltip(defaultCombination.value, tooltip)
        combination.addListener { _: ObservableValue<out KeyCodeCombination>, oldValue: KeyCodeCombination?, newValue: KeyCodeCombination ->
            if (null == oldValue) {
                updateTooltip(newValue, tooltip)
            }
            text = newValue.displayText
        }
        activated.addListener { _: ObservableValue<out Boolean>, _: Boolean, newValue: Boolean ->
            text = if (newValue) pressKey else combination.value!!.displayText
        }
        isEditable = false
        onKeyReleased = EventHandler {
            if (activated.value) {
                if (it.code.isLetterKey || it.code.isDigitKey) combination.value = toCombination(it)
                activated.value = false
            }
        }
        focusedProperty().addListener { _: ObservableValue<out Boolean>, _: Boolean, newValue: Boolean ->
            activated.setValue(newValue)
        }
        onMouseClicked = EventHandler {
            requestFocus()
            activated.setValue(true)
        }
    }

    fun activatedProperty(): ReadOnlyProperty<Boolean> {
        return activated
    }

    fun combinationProperty(): ReadOnlyProperty<KeyCodeCombination> {
        return combination
    }

    fun defaultCombinationProperty(): Property<KeyCodeCombination> {
        return defaultCombination
    }

    fun getActivated(): Boolean {
        return activated.value
    }

    fun getCombination(): KeyCodeCombination {
        return combination.value
    }

    fun getDefaultCombination(): KeyCodeCombination {
        return defaultCombination.value
    }

    fun setDefaultCombination(defaultCombination: KeyCodeCombination) {
        this.defaultCombination.value = defaultCombination
    }

    fun reset() {
        combination.value = defaultCombination.value
        activated.value = false
    }

    private fun updateTooltip(keyCodeCombination: KeyCodeCombination, tooltipString: String) {
        FxRun.runLater {
            tooltip = Tooltip("$tooltipString ${keyCodeCombination.displayText}")
        }
    }

    companion object {
        val NONE: KeyCodeCombination = KeyCodeCombination(KeyCode.CODE_INPUT)
        private const val LC_DEFAULT = "default-is"
        private const val LC_PRESS_ANY_KEY = "press-any-key"
        private fun toCombination(event: KeyEvent): KeyCodeCombination =
            KeyCodeCombination(
                event.code,
                if (event.isShiftDown) KeyCombination.ModifierValue.DOWN else KeyCombination.ModifierValue.UP,
                if (event.isControlDown) KeyCombination.ModifierValue.DOWN else KeyCombination.ModifierValue.UP,
                if (event.isAltDown) KeyCombination.ModifierValue.DOWN else KeyCombination.ModifierValue.UP,
                if (event.isMetaDown) KeyCombination.ModifierValue.DOWN else KeyCombination.ModifierValue.UP,
                if (event.isShortcutDown) KeyCombination.ModifierValue.DOWN else KeyCombination.ModifierValue.UP
            )
    }
}
